tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs

```diff
@@ -212,6 +212,46 @@ public void Dispose_should_raise_the_correct_events()
             _capturedEvents.Any().Should().BeFalse();
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void ReceiveMessage_should_throw_a_FormatException_when_message_is_an_invalid_size(
+            [Values(-1, 48000001)]
+            int length,
+            [Values(false, true)]
+            bool async)
+        {
+            using (var stream = new BlockingMemoryStream())
+            {
+                var bytes = BitConverter.GetBytes(length);
+                stream.Write(bytes, 0, bytes.Length);
+                stream.Seek(0, SeekOrigin.Begin);
+                var encoderSelector = new ReplyMessageEncoderSelector<BsonDocument>(BsonDocumentSerializer.Instance);
+
+                Exception exception;
+                if (async)
+                {
+                    _mockStreamFactory.Setup(f => f.CreateStreamAsync(_endPoint, CancellationToken.None))
```

Comment:
It can be written in this way which probably looks better:

                    _mockStreamFactory
                        .Setup(f => f.CreateStreamAsync(_endPoint, CancellationToken.None))
                        .ReturnsAsync(stream);
                    _subject.OpenAsync(CancellationToken.None).GetAwaiter().GetResult();
                    exception = Record
                        .ExceptionAsync(async () => await _subject.ReceiveMessageAsync(10, encoderSelector, _messageEncoderSettings, CancellationToken.None))
                        .Result;

Reply:
`ExceptionAsync` is another method I did not know about. 

I'm not sure about this one though... 

I don't think we can use `.Result`. We would have to use `.GetAwaiter().GetResult()` to unwrap exceptions properly.

Reply:
If we chose to use Xunit's support for `async Task` test methods we could probably use this:

        exception = await Record.ExceptionAsync(() => _subject.ReceiveMessageAsync(10, encoderSelector, _messageEncoderSettings, CancellationToken.None));

Reply:
One more new feature for me :) 
I didn't know that `async Task` tests are supported. So probably we need to write tests which include `async` methods in a similar way to this one:
       
        [Theory]
        [ParameterAttributeData]
        public async Task ReceiveMessage_should_throw_a_FormatException_when_message_is_an_invalid_size(
            [Values(-1, 48000001)] int length,
            [Values(false, true)] bool async)

It allows us just to use `await` rather than `GetAvaiter` and so on. Probably we need to even add it into our a google doc with code format questions? Thoughts?

Reply:
This was discussed here:

https://github.com/ksadoff/mongo-csharp-driver/pull/6#discussion_r296410104

I certainly wouldn't go changing any existing tests.

I agree that using `async Task` test methods looks attractive, but I'm not 100% sure whether there are any negative side effects from blocking (when calling sync methods) in an async method.

Reply:
If you decide to change this new test to be an async test method be sure to use `async Task` and not `async void`.

@vincentkam I did not catch this when you first introduced the two existing tests that use `async void`.

Reply:
I'm finding hits on Google that blocking in an async Xunit test method can lead to deadlocks.

Perhaps we should stick to what we've been doing all along, and use sync test methods combined with `GetAwaiter().GetResult()` when calling the async method under test.

Reply:
I'm also finding comments that using async Xunit test methods can lead to an unbounded number of tests running in parallel.

Apparently every time a test method hits an await statement it is possible for Xunit to schedule a new test method, even though the current test has not finished.

From the limited understanding I have this can't happen with sync test methods since there is a fixed size thread pool to run sync tests on, so the max number of tests running at the same time is limited by the size of the underlying thread pool.

Nice as this new async Xunit test method feature is I'm getting nervous about using it. Might be better to stick to what's been working.

Reply:
> If you decide to change this new test to be an async test method be sure to use `async Task` and not `async void`.
> 
> @vincentkam I did not catch this when you first introduced the two existing tests that use `async void`.

Ah you're right, `async Task` would be preferable I think. Fwiw, I vaguely recall that XUnit has special machinery for handling `async void` to [call and capture exceptions](https://stackoverflow.com/questions/23824660/xunit-async-test-not-working-properly).

Reply:
>Might be better to stick to what's been working.
based on possible problems that @rstam mentioned above(and I didn't google any proofs that using `async` and `sync` together is safe), I agree, that probably it's better to use `sync` test methods.

Reply:
>I don't think we can use .Result. We would have to use .GetAwaiter().GetResult() to unwrap exceptions properly.
Agree about it.